cmd/kube-controller-manager/app/core.go

```diff
@@ -586,35 +576,44 @@ func processCIDRs(cidrsList string) ([]*net.IPNet, bool, error) {
 }
 
 // setNodeCIDRMaskSizes returns the IPv4 and IPv6 node cidr mask sizes.
-// If --node-cidr-mask-size not set, then it will return default IPv4 and IPv6 cidr mask sizes.
-func setNodeCIDRMaskSizes(cfg nodeipamconfig.NodeIPAMControllerConfiguration) (int, int, error) {
+func setNodeCIDRMaskSizes(clusterCIDRs []*net.IPNet, cfg nodeipamconfig.NodeIPAMControllerConfiguration) (int, int, error) {
 	ipv4Mask, ipv6Mask := defaultNodeMaskCIDRIPv4, defaultNodeMaskCIDRIPv6
-	// NodeCIDRMaskSizeIPv4 and NodeCIDRMaskSizeIPv6 can be used only for dual-stack clusters
-	if cfg.NodeCIDRMaskSizeIPv4 != 0 || cfg.NodeCIDRMaskSizeIPv6 != 0 {
-		return ipv4Mask, ipv6Mask, errors.New("usage of --node-cidr-mask-size-ipv4 and --node-cidr-mask-size-ipv6 are not allowed with non dual-stack clusters")
-	}
-	if cfg.NodeCIDRMaskSize != 0 {
-		ipv4Mask = int(cfg.NodeCIDRMaskSize)
-		ipv6Mask = int(cfg.NodeCIDRMaskSize)
-	}
-	return ipv4Mask, ipv6Mask, nil
-}
 
-// setNodeCIDRMaskSizesDualStack returns the IPv4 and IPv6 node cidr mask sizes to the value provided
-// for --node-cidr-mask-size-ipv4 and --node-cidr-mask-size-ipv6 respectively. If value not provided,
-// then it will return default IPv4 and IPv6 cidr mask sizes.
-func setNodeCIDRMaskSizesDualStack(cfg nodeipamconfig.NodeIPAMControllerConfiguration) (int, int, error) {
-	ipv4Mask, ipv6Mask := defaultNodeMaskCIDRIPv4, defaultNodeMaskCIDRIPv6
-	// NodeCIDRMaskSize can be used only for single stack clusters
-	if cfg.NodeCIDRMaskSize != 0 {
-		return ipv4Mask, ipv6Mask, errors.New("usage of --node-cidr-mask-size is not allowed with dual-stack clusters")
-	}
-	if cfg.NodeCIDRMaskSizeIPv4 != 0 {
-		ipv4Mask = int(cfg.NodeCIDRMaskSizeIPv4)
-	}
-	if cfg.NodeCIDRMaskSizeIPv6 != 0 {
-		ipv6Mask = int(cfg.NodeCIDRMaskSizeIPv6)
+	switch len(clusterCIDRs) {
+	case 1:
+		if netutils.IsIPv6CIDR(clusterCIDRs[0]) {
```

Comment:
so actually at this point maybe it would make sense to merge `setNodeCIDRMaskSizes` and `getNodeCIDRMaskSizes` into a single function?

anyway, basically lgtm other than maybe that, and the issue of specifying redundant/contradictory args

Reply:
> so actually at this point maybe it would make sense to merge `setNodeCIDRMaskSizes` and `getNodeCIDRMaskSizes` into a single function?
> 
> anyway, basically lgtm other than maybe that, and the issue of specifying redundant/contradictory args

Done.